tests/nn/data_parallel/test_fsdp.py

```diff
@@ -697,6 +697,38 @@ def _test_no_sync(self, model, batch_dim):
         assert objects_are_equal(ref_grads, accumulated_grads, raise_exception=True)
 
 
+class TestAutoWrap(unittest.TestCase):
+    def setUp(self) -> None:
+        class TestDistributedGroup:
+            def size(self):
+                return 1
+
+            @property
+            def world_size(self):
+                return 1
+
+            def rank(self):
+                return 0
+
+        self.process_group = TestDistributedGroup()
+
+    def test_auto_wrap(self):
+        with FullyShardedDataParallel.enable_auto_wrap(flatten_parameters=False, process_group=self.process_group):
+            layer = FullyShardedDataParallel.auto_wrap(torch.nn.Linear(5, 5))
+        assert isinstance(layer, FullyShardedDataParallel)
+        assert layer.flatten_parameters is False
+
+    def test_auto_wrap_disabled_outside_context(self):
+        layer = FullyShardedDataParallel.auto_wrap(torch.nn.Linear(5, 5))
+        assert isinstance(layer, torch.nn.Linear)
+
+    def test_auto_wrap_override_defaults(self):
```

Comment:
also, is it worth testing nested wrapping like:

with FSDP.config_context()
    FSDP.wrap_if_in...
        with FSDP.config_context()
               FSDP.wrap_if_in

Reply:
I'm not sure if we want to support nested configuration contexts, I could imagine it getting tricky to resolve conflicting options.

One exception is we may want to allow certain layers to override a limited set of options, e.g., MoE layers will use a dummy process group of size 1 to turn off sharding. Possibly some layers would want to turn off mixed precision. I suggest we can have this limited set of overrides be directly part of the "auto_wrap" (new name TBD) call. Something like: `FSDP.auto_wrap(moe_layer, process_group=custom_process_group)` or something like that.

What do you think?